### ISSUE TITLE
Function arguments in trait functions are "unused" if unused in default impl

### ISSUE THREAD

**Issue by birkenfeld** [rust-lang/rust#55070]
E.g.
```
trait Foo {
    fn frob(id: usize) -> Option<usize> { None }
}
```
triggers "unused variable: id".

I would argue that this shouldn't trigger the lint, since it isn't triggered either when there is no default impl.  The argument name has two roles here, one of which the lint doesn't consider.

**Comment by sfackler** [comment]
Duplicate of #26487.

**Comment by estebank** [comment]
Current output:

```
warning: unused variable: `id`
 --> file5.rs:2:13
  |
2 |     fn frob(id: usize) -> Option<usize> { None }
  |             ^^ help: consider prefixing with an underscore: `_id`
  |
  = note: `#[warn(unused_variables)]` on by default
```

As per https://github.com/rust-lang/rust/issues/26487#issuecomment-169828205, this is not something we would like to change, but I'll let @rust-lang/lang decide whether that is still the case and close, or prioritize.

**Comment by Centril** [comment]
I'm a bit conflicted but ultimately I agree with the prior consensus. It seems to me nothing has changed from before and that `#[allow(...)]` (enhanced by https://github.com/rust-lang/rust/issues/60406) or `let _ = id;`  works.
